Weird behavior with ForeignKey models

I came across a weird behavior with trying to use `ForeignKey` models in migrations. I pushed a commit (https://github.com/percipient/django_migration_testcase/commit/9a3d3242e9424e09711da9d2168845e66909eb8f) that illustrates it with a failing test.

Pretty much if you have a `ForeignKey` and attempt to create a new instance of a model (after using `get_model_before` or `get_model_after`), you get strange errors about the objects not being the correct type to add as a FK.

The results of running the added tests in that commit (I only ran SQLite, I don't have PostgreSQL installed):

```
Tests for tests/test_project_sqlite_django
Creating test database for alias 'default'...
...E.F...
======================================================================
ERROR: test_migration (test_app.tests.ForeignKeyTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "django_migration_testcase/tests/test_project_sqlite_django/test_app/tests.py", line 134, in test_migration
    foreign = ForeignModel(name='test_foreign', my=my)
  File "~/.virtualenvs/django_migration_testcase/lib/python2.7/site-packages/django/db/models/base.py", line 468, in __init__
    setattr(self, field.name, rel_obj)
  File "~/.virtualenvs/django_migration_testcase/lib/python2.7/site-packages/django/db/models/fields/related.py", line 635, in __set__
    self.field.rel.to._meta.object_name,
ValueError: Cannot assign "<MyModel: MyModel object>": "ForeignModel.my" must be a "MyModel" instance.

======================================================================
FAIL: test_migration_clearly (test_app.tests.ForeignKeyTest)
A clear illustration of the problem.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "django_migration_testcase/tests/test_project_sqlite_django/test_app/tests.py", line 167, in test_migration_clearly
    self.assertEqual(MyModel, MyModel2)
AssertionError: <class 'MyModel'> != <class 'MyModel'>

----------------------------------------------------------------------
Ran 9 tests in 1.449s

FAILED (failures=1, errors=1)
```

I've tried printing out additional information in `test_migration_clearly`, but as far as I can tell they *are* the same class.

I'm not 100% sure this is a bug, maybe I'm just doing something silly, but I've spent a bit of time on this and have been unable to figure it out. Any help would be appreciated!

--- a/django_migration_testcase/django_migrations.py
+++ b/django_migration_testcase/django_migrations.py
@@ -10,6 +10,9 @@ class MigrationTest(TransactionTestCase):
     __abstract__ = True
 
     def setUp(self):
+        self.apps_before = None
+        self.apps_after = None
+
         super(MigrationTest, self).setUp()
         call_command('migrate', self.app_name, self.before,
                      no_initial_data=True, verbosity=0)
@@ -36,13 +39,15 @@ class MigrationTest(TransactionTestCase):
 
     def get_model_before(self, model_name):
         app_name, model_name = self._get_app_and_model_name(model_name)
-        return (self._get_apps_for_migration(app_name, self.before)
-                .get_model(app_name, model_name))
+        if not self.apps_before:
+            self.apps_before = self._get_apps_for_migration(app_name, self.before)
+        return self.apps_before.get_model(app_name, model_name)
 
     def get_model_after(self, model_name):
         app_name, model_name = self._get_app_and_model_name(model_name)
-        return (self._get_apps_for_migration(app_name, self.after)
-                .get_model(app_name, model_name))
+        if not self.apps_after:
+            self.apps_after = self._get_apps_for_migration(app_name, self.after)
+        return self.apps_after.get_model(app_name, model_name)
 
     def run_migration(self):
         call_command('migrate', self.app_name, self.after,

--- a/tests/test_app/migrations/0005_foreignmodel.py
+++ b/tests/test_app/migrations/0005_foreignmodel.py
@@ -1,0 +1,22 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.db import models, migrations
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('test_app', '0004_populate_mymodel_double_number'),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='ForeignModel',
+            fields=[
+                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
+                ('name', models.CharField(max_length=100)),
+                ('my', models.ForeignKey(to='test_app.MyModel')),
+            ],
+        ),
+    ]

--- a/tests/test_app/models.py
+++ b/tests/test_app/models.py
@@ -5,3 +5,8 @@ class MyModel(models.Model):
     name = models.CharField(max_length=100)
     number = models.IntegerField(null=True)
     double_number = models.IntegerField(null=True)
+
+
+class ForeignModel(models.Model):
+    name = models.CharField(max_length=100)
+    my = models.ForeignKey(MyModel)

--- a/tests/test_app/south_migrations/0005_auto__add_foreignmodel.py
+++ b/tests/test_app/south_migrations/0005_auto__add_foreignmodel.py
@@ -1,0 +1,37 @@
+# -*- coding: utf-8 -*-
+from south.db import db
+from south.v2 import SchemaMigration
+
+
+class Migration(SchemaMigration):
+
+    def forwards(self, orm):
+        # Adding model 'ForeignModel'
+        db.create_table(u'test_app_foreignmodel', (
+            (u'id', self.gf('django.db.models.fields.AutoField')(primary_key=True)),
+            ('name', self.gf('django.db.models.fields.CharField')(max_length=100)),
+            ('my', self.gf('django.db.models.fields.related.ForeignKey')(to=orm['test_app.MyModel'])),
+        ))
+        db.send_create_signal(u'test_app', ['ForeignModel'])
+
+    def backwards(self, orm):
+        # Deleting model 'ForeignModel'
+        db.delete_table(u'test_app_foreignmodel')
+
+    models = {
+        u'test_app.foreignmodel': {
+            'Meta': {'object_name': 'ForeignModel'},
+            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
+            'my': ('django.db.models.fields.related.ForeignKey', [], {'to': u"orm['test_app.MyModel']"}),
+            'name': ('django.db.models.fields.CharField', [], {'max_length': '100'})
+        },
+        u'test_app.mymodel': {
+            'Meta': {'object_name': 'MyModel'},
+            'double_number': ('django.db.models.fields.IntegerField', [], {'null': 'True'}),
+            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
+            'name': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
+            'number': ('django.db.models.fields.IntegerField', [], {'null': 'True'})
+        }
+    }
+
+    complete_apps = ['test_app']

--- a/tests/test_app/tests.py
+++ b/tests/test_app/tests.py
@@ -129,7 +129,7 @@ class ForeignKeyTest(MigrationTest):
         my = MyModel(name='test_my', number=1, double_number=3.14)
         my.save()
 
-        foreign = ForeignModel(name='test_foreign', my=my)
+        ForeignModel(name='test_foreign', my=my)
 
     def test_migration2(self):
         """Same test as test_migration, but this one passes."""
@@ -149,7 +149,7 @@ class ForeignKeyTest(MigrationTest):
         my = MyModel(name='test_my', number=1, double_number=3.14)
         my.save()
 
-        foreign = ForeignModel(name='test_foreign', my=my)
+        ForeignModel(name='test_foreign', my=my)
 
     def test_migration_clearly(self):
         """A clear illustration of the problem."""

--- a/tests/test_app/tests.py
+++ b/tests/test_app/tests.py
@@ -107,3 +107,59 @@ class GetModelMigrationTest(MigrationTest):
 
         MyModel = self.get_model_after('test_app.MyModel')
         self.assertEqual(MyModel.__name__, 'MyModel')
+
+
+class ForeignKeyTest(MigrationTest):
+    before = '0004_populate_mymodel_double_number'
+    after = '0005_foreignmodel'
+    app_name = 'test_app'
+
+    def test_migration(self):
+        MyModel = self.get_model_before('test_app.MyModel')
+        self.assertEqual(MyModel.__name__, 'MyModel')
+
+        self.run_migration()
+
+        ForeignModel = self.get_model_after('test_app.ForeignModel')
+        self.assertEqual(ForeignModel.__name__, 'ForeignModel')
+
+        MyModel = self.get_model_after('test_app.MyModel')
+        self.assertEqual(MyModel.__name__, 'MyModel')
+
+        my = MyModel(name='test_my', number=1, double_number=3.14)
+        my.save()
+
+        foreign = ForeignModel(name='test_foreign', my=my)
+
+    def test_migration2(self):
+        """Same test as test_migration, but this one passes."""
+        MyModel = self.get_model_before('test_app.MyModel')
+        self.assertEqual(MyModel.__name__, 'MyModel')
+
+        self.run_migration()
+
+        ForeignModel = self.get_model_after('test_app.ForeignModel')
+        self.assertEqual(ForeignModel.__name__, 'ForeignModel')
+
+        # get_model_before/get_model_after seems to not get the same model as
+        # this crazy thing.
+        MyModel = ForeignModel.my.field.rel.to
+        self.assertEqual(MyModel.__name__, 'MyModel')
+
+        my = MyModel(name='test_my', number=1, double_number=3.14)
+        my.save()
+
+        foreign = ForeignModel(name='test_foreign', my=my)
+
+    def test_migration_clearly(self):
+        """A clear illustration of the problem."""
+        self.run_migration()
+
+        ForeignModel = self.get_model_after('test_app.ForeignModel')
+
+        # get_model_before/get_model_after seems to not get the same model as
+        # this crazy thing.
+        MyModel = ForeignModel.my.field.rel.to
+        MyModel2 = self.get_model_after('test_app.MyModel')
+
+        self.assertEqual(MyModel, MyModel2)

